In [47]:
import pandas as pd
import numpy as np 

In [23]:
pd.__version__

'2.3.1'

### Question 1 
#### Pandas version: <u>2.3.1</u>

In [24]:
df = pd.read_csv("car_fuel_efficiency.csv")

In [25]:
df.shape

(9704, 11)

### Question 2
#### Records Count: <u>9704</u>


In [26]:
df.columns

Index(['engine_displacement', 'num_cylinders', 'horsepower', 'vehicle_weight',
       'acceleration', 'model_year', 'origin', 'fuel_type', 'drivetrain',
       'num_doors', 'fuel_efficiency_mpg'],
      dtype='object')

In [27]:
df.fuel_type.unique()

array(['Gasoline', 'Diesel'], dtype=object)

In [28]:
df['fuel_type'].nunique()

2

### Question 3
#### Fuel Types: <u>2</u>

In [29]:
df.isnull().sum()

engine_displacement      0
num_cylinders          482
horsepower             708
vehicle_weight           0
acceleration           930
model_year               0
origin                   0
fuel_type                0
drivetrain               0
num_doors              502
fuel_efficiency_mpg      0
dtype: int64

### Question 4
#### Missing values (num of columns)
##### 4 columns have missing values:
<ul>
<li>num_cylinders</li>
<li>horsepower</li>
<li>acceleration</li>
<li>num_doors</li>
</ul>

In [30]:
df[df['origin']=='Asia']['fuel_efficiency_mpg'].max().round(2)

np.float64(23.76)

### Question 5
#### Max fuel efficiency of Asian cars (mpg): <u>23.76 mpg</u>

### Q6. Median value of horsepower steps
<ol>
<li>Find the median value of horsepower column in the dataset.</li>
<li>Next, calculate the most frequent value of the same horsepower column.</li>
<li>Use fillna method to fill the missing values in horsepower column with the most frequent value from the previous step.</li>
<li>Now, calculate the median value of horsepower once again.</li>

In [31]:
#1 Find median 
df['horsepower'].median()

np.float64(149.0)

In [33]:
#2 find mode
# df['horsepower'].mode()[0]
mode_value = df['horsepower'].mode()[0]
mode_value

np.float64(152.0)

In [34]:
#3 fillna with mode, #4 get new median
df['horsepower'] = df['horsepower'].fillna(mode_value)
df['horsepower'].median()

np.float64(152.0)

### Question 6: Did median change? 
#### Yes, increased (149 -> 152)

### Q7. Sum of weights
<ol>
<li>Select all the cars from Asia
<li>Select only columns vehicle_weight and model_year
<li>Select the first 7 values
<li>Get the underlying NumPy array. Let's call it X.
<li>Compute matrix-matrix multiplication between the transpose of X and X. To get the transpose, use X.T. <li>Let's call the result XTX.
<li>Invert XTX.
<li>Create an array y with values [1100, 1300, 800, 900, 1000, 1100, 1200].
<li>Multiply the inverse of XTX with the transpose of X, and then multiply the result by y. Call the result w.
<li>What's the sum of all the elements of the result?

In [37]:
#1 select all cars of Asia 
df_asian = df[df['origin']=='Asia']

In [38]:
#2 Select only columns vehicle_weight and model_year
df_asian_sub = df_asian[['vehicle_weight', 'model_year']]

In [ ]:
# 3 get first 7 rows
df_asian_sub_seven = df_asian_sub.iloc[0:7]

In [42]:
df_asian_sub_seven

vehicle_weight  model_year
8      2714.219310        2016
12     2783.868974        2010
14     3582.687368        2007
20     2231.808142        2011
21     2659.431451        2016
34     2844.227534        2014
38     3761.994038        2019

In [43]:
# 4 get numpy array 
X = df_asian_sub_seven.to_numpy()

In [44]:
type(X)

numpy.ndarray

In [50]:
# 5 Compute matrix-matrix multiplication between the transpose of X and X. To get the transpose, use X.T. Let's call the result XTX
XTX = X.T.dot(X)
XTX

array([[62248334.33150762, 41431216.5073268 ],
       [41431216.5073268 , 28373339.        ]])

In [51]:
# 6 invert 
XTX_inv = np.linalg.inv(XTX)
XTX_inv

array([[ 5.71497081e-07, -8.34509443e-07],
       [-8.34509443e-07,  1.25380877e-06]])

In [53]:
# 7 Create an array y with values [1100, 1300, 800, 900, 1000, 1100, 1200].
y = np.array([1100, 1300, 800, 900, 1000, 1100, 1200])
y

array([1100, 1300,  800,  900, 1000, 1100, 1200])

In [54]:
# calculate w 
w = XTX_inv.dot(X.T).dot(y)
w

array([0.01386421, 0.5049067 ])

In [57]:
# sum of w elements 
w.sum().round(3)

np.float64(0.519)